RecoParticleFlow/PFClusterProducer/src/PFMultiDepthClusterizer.cc

```diff
@@ -97,54 +95,59 @@ buildClusters(const reco::PFClusterCollection& input,
 
 
 void PFMultiDepthClusterizer::
-calculateShowerShapes(const reco::PFClusterCollection& clusters,std::vector<double>& etaRMS,std::vector<double>& phiRMS) {
-  float etaSum;
-  float phiSum;
+calculateShowerShapes(const reco::PFClusterCollection& clusters,std::vector<double>& etaRMS2,std::vector<double>& phiRMS2) {
 
   //shower shapes. here do not use the fractions 
 
   for( unsigned int i=0;i<clusters.size();++i ) {
     const reco::PFCluster& cluster = clusters[i]; 
-    etaSum=0.0;phiSum=0.0;
+    double etaSum=0.0; double phiSum=0.0;
+    auto const & crep = cluster.positionREP();
     for (const auto& frac : cluster.recHitFractions()) {
-      etaSum +=frac.fraction()*frac.recHitRef()->energy()*fabs(frac.recHitRef()->position().Eta() -cluster.position().Eta() );
-      phiSum +=frac.fraction()*frac.recHitRef()->energy()*fabs(deltaPhi(frac.recHitRef()->position().Phi(),cluster.position().Phi()));
+      auto const & h = *frac.recHitRef();
+      auto const & rep = h.positionREP();
+      etaSum += (frac.fraction()*h.energy())*std::abs(rep.eta()-crep.eta());
+      phiSum += (frac.fraction()*h.energy())*std::abs(deltaPhi(rep.phi(),crep.phi()));
     }
     //protection for single line : assign ~ tower
-    etaRMS[i] = std::max(etaSum/cluster.energy(),0.1);
-    phiRMS[i] = std::max(phiSum/cluster.energy(),0.1);
-    
+    etaRMS2[i] = std::max(etaSum/cluster.energy(),0.1);
+    etaRMS2[i]*= etaRMS2[i];
+    phiRMS2[i] = std::max(phiSum/cluster.energy(),0.1);
+    phiRMS2[i]*=phiRMS2[i];
   }
 
 }
 
 
 std::vector<PFMultiDepthClusterizer::ClusterLink>   PFMultiDepthClusterizer::
-link(const reco::PFClusterCollection& clusters ,const std::vector<double>& etaRMS,const std::vector<double>& phiRMS) {
+link(const reco::PFClusterCollection& clusters ,const std::vector<double>& etaRMS2,const std::vector<double>& phiRMS2) {
 
   std::vector<ClusterLink> links;
   //loop on all pairs
   for (unsigned int i=0;i<clusters.size();++i)
     for (unsigned int j=0;j<clusters.size();++j) {
-      if (i==j )
-	continue;
+      if (i==j )  continue;
 
       const reco::PFCluster& cluster1 = clusters[i]; 
       const reco::PFCluster& cluster2 = clusters[j]; 
 
-      float dz = (cluster2.depth() - cluster1.depth());
+      auto dz = (cluster2.depth() - cluster1.depth());
 
       //Do not link at the same layer and only link inside out!
-      if (dz<0.0 || fabs(dz)<0.2)
-	continue;
-
-      float deta =(cluster1.position().Eta()-cluster2.position().Eta())*(cluster1.position().Eta()-cluster2.position().Eta())/(etaRMS[i]*etaRMS[i]+etaRMS[j]*etaRMS[j]);
-      float dphi = deltaPhi(cluster1.position().Phi(),cluster2.position().Phi())*deltaPhi(cluster1.position().Phi(),cluster2.position().Phi())/(phiRMS[i]*phiRMS[i]+phiRMS[j]*phiRMS[j]);
+      if (dz<0.0f || std::abs(dz)<0.2f)  continue;
+ 
+      auto const & crep1 = cluster1.positionREP();
+      auto const & crep2 = cluster2.positionREP();
+ 
+      auto deta = crep1.eta()-crep2.eta();
```

Comment:
the regression is caused by the bizarre mixed FP commutations in the original code.
Here double are used consistently.
In future we can migrate to float ...